[RNMobile] Fix isURL

etoledom opened the pull request:
`gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1879

## Description
<!-- Please describe what you have changed or added -->
The changes on the PR #19871 to `isURL` assume that the constructor of `URL` "throws an error for an invalid URL", which is not the case for the [mobile implementation.](https://github.com/facebook/react-native/blob/6ba2aeefa8dfe031bf1dc46dbea29235aec31d61/Libraries/Blob/URL.js#L133-L137)

This PR brings back the previous implementation of `isURL` just for native, with a few improvements to make it behave as the new web version (tested running the same set of tests).

@SergioEstevao - I remember you having some opinions about the `isURL` check, could you please take a look at this?

Also @aduth - since you authored the original PR, it would be great if we could have a shared implementation on this, but from what I see, we can not trust that the URL constructor will throw.

Any improvement to the given solution is welcome!
The implemented Regex was taken from [here](https://urlregex.com), adding a group to match for ports.

## How has this been tested?
- The new native Unit Tests should pass.
- Test on the related [gutenberg-mobile](https://github.com/wordpress-mobile/gutenberg-mobile/pull/1879) PR following the given test steps.

etoledom committed:
Fix mobile version of is-url.
Changes on this PR: https://github.com/WordPress/gutenberg/pull/19871 broke the native `is-url` version since it won't throw if the given string is not a url.

etoledom committed:
Mobile unit tests to is-url + better REGEXP

etoledom committed:
Fix lint issues

aduth commented:
> Also @aduth - since you authored the original PR, it would be great if we could have a shared implementation on this, but from what I see, we can not trust that the URL constructor will throw.

In putting together #19871, I was careful to consider that validation conform to the contract of what is considered valid per the URL standard, and only incidentally that this be fulfilled using the `URL` constructor. For how it impacts here, I agree that it would be nice if the implementation would be shared, but strictly speaking it does not need to.

That said:

- I feel that React Native's implementation of `URL` is not spec-conformant. The specification [dictates that "if parsedURL is failure, then throw a `TypeError`"](https://url.spec.whatwg.org/#dom-url-url). I observe that the React native implementation of `URLSearchParams` [includes some parts of `jsdom/whatwg-url`](https://github.com/facebook/react-native/blob/6ba2aeefa8dfe031bf1dc46dbea29235aec31d61/Libraries/Blob/URL.js#L54), where that package's implementation of `URL` [does throw](https://github.com/jsdom/whatwg-url/blob/c12cc618ce5c745bf53a1a845701f0486be2aa2a/lib/URL-impl.js). This doesn't impact how we decide to proceed for the short-term, but it might be something to report for upstream improvement.
- Would it be an option for the native implementation to use one of either the [`jsdom/whatwg-url`](https://github.com/jsdom/whatwg-url) or [the URL polyfill we use for unsupported browsers](https://github.com/Financial-Times/polyfill-library/blob/master/polyfills/URL/polyfill.js) instead, at least to be equally spec-conforming?
- I can grant that the changes in this pull request would at least be the minimal changes necessary to restore some workable behavior, but I would not be comfortable with the diverging expected outputs as a long-term solution.

aduth commented:
> * Would it be an option for the native implementation to use one of either the [`jsdom/whatwg-url`](https://github.com/jsdom/whatwg-url) or [the URL polyfill we use for unsupported browsers](https://github.com/Financial-Times/polyfill-library/blob/master/polyfills/URL/polyfill.js) instead, at least to be equally spec-conforming?

I'm glad I brought this up, because in closer inspection of the polyfill we chose to use for `URL`, it also does not appear to throw errors as we are relying on it to do for URL-checking.

aduth commented:
See also: https://github.com/charpeni/react-native-url-polyfill

Specifically: https://github.com/charpeni/react-native-url-polyfill#why-do-we-need-this

etoledom commented:
Thank you @aduth for your feedback! I've tested using `react-native-url-polyfill` and it seems to do the job! 🎉 

Updated the PR with this change. I also added this package as deb dependency to gutenberg to run the native unit tests properly.

aduth reviewed (approved):
Looks sensible from my perspective 👍 

I'll leave it to your discretion whether it needs eyes from one more familiar with mobile side of things.

aduth reviewed (changes requested):
Actually, can you confirm: Should `react-native-url-polyfill` be defined as a dependency of the `url` package rather than in the root `package.json`? Typically I would expect to see a package's own dependencies listed only in its own `package.json`, not at the root. Unsure if this is different for React Native work.

etoledom commented:
@aduth - some updates: `react-native-url-polyfill` is giving us troubles on Android with the current setup. It crashes with this error:

![Screenshot 2020-02-12 at 19 07 56](https://user-images.githubusercontent.com/9772967/74415593-f9444e00-4e43-11ea-901a-901249291fb6.png)

Talking with @hypest internally, we have decided to go back to the first proposal momentarily, in order to fix https://github.com/wordpress-mobile/gutenberg-mobile/issues/1875, and then we would come back to `react-native-url-polyfill` as soon as the Android issue is solved.

I'll wait for your confirmation before to roll back the commits 👍

aduth commented:
> then we would come back to `react-native-url-polyfill` as soon as the Android issue is solved.

Is this a known upstream issue? Do we have a sense why the error is happening?

If it's simplest to revert to the first proposal as a fix for https://github.com/wordpress-mobile/gutenberg-mobile/issues/1875, that sounds reasonable to me. I would just be wary to not let these diverge too long, since the two implementations can have differing resulting values from its input (the earlier implementation being much more strict about "http" URLs).

etoledom committed:
Remove not-needed comment

etoledom commented:
Reverted to the previous proposal. 

> the earlier implementation being much more strict about "http" URLs.

@aduth - This proposal is a bit more complex than the older implementation. I was careful to make the[ same set of tests](https://github.com/WordPress/gutenberg/blob/6512ab834dcf90947f4988c2c79f3cacb8640137/packages/url/src/test/index.native.js) implemented on the web side pass. So things like:
```
[ 'http://wordpress.org' ],
[ 'https://wordpress.org' ],
[ 'mailto:example@example.com' ],
[ 'ssh://user:password@127.0.0.1:8080' ],
```
...will all return true. I hope this gives a bit more confidence. But it's true that there might be other urls that these implementations will return different results.

There is this [related issue](https://github.com/facebook/react-native/issues/26645), I think it's the same problem but haven't had much attention so far. We might need to try to push it a bit.

aduth commented on packages/url/src/is-url.native.js:
```diff
@@ -0,0 +1,18 @@
+ */
+export function isURL( url ) {
+	const match = url.match( URL_REGEXP );
+	return match !== null && match.length >= 1 && match[ 0 ] === url;
```
It was a bit confusing for me to understand why it wasn't enough to implement this as simply:

```js
return URL_REGEXP.test( url );
```

Running it locally, I can see it doesn't work, and it seems like the pattern matches a lot of "invalid" URLs that must be subjected to further conditions. It would help if there were an inline code comment explaining the logic of what is happening here.

etoledom replied:
Added on https://github.com/WordPress/gutenberg/pull/20172/commits/c16b34e959914e2bacc3e0daa1db52f0a5641104 👍

aduth reviewed (approved):
If it can pass the same tests, that seems like a perfectly reasonable alternative implementation. And if it turns out not to be, then it's a clear indication that the tests could be improved. A win/win either way! (Well, except for those affected by however it might hypothetically break 😅)

Noting that this aligns to a remark I made at https://github.com/WordPress/gutenberg/pull/19871#discussion_r370775828 in the original implementation:

>One advantage of having the tests is that it could support some potential future refactoring where we choose not to implement it using the `URL` constructor.

etoledom committed:
Adding inline comments to explain `.match()` decision over `.test()`.

aduth commented on packages/url/src/test/index.native.js:
```diff
@@ -0,0 +1,33 @@
+/**
```
Would it work just as well if this whole file was with:

```js
import './index.js';
```

?

etoledom replied:
Good call, and it works! 🎉 
Best part is all tests are ✅ 

https://github.com/WordPress/gutenberg/pull/20172/commits/fa15b578c2598f2d4b4a02821e22bf32cc3932e6

Tug replied:
Oh nice we share tests with web, ignore my comment :)

etoledom committed:
Sharing test suit with web.

Tug reviewed (approved):
LGTM 👍

etoledom commented:
Thank you both! 🎉

charpeni commented:
👋 Hey!

I'm the author of [react-native-url-polyfill](https://github.com/charpeni/react-native-url-polyfill). I flagged the issue with Hermes: https://github.com/facebook/hermes/issues/193, and I fixed it in react-native-url-polyfill. Everything is now working fine with Hermes!

aduth commented:
Hi @charpeni , thanks for the heads-up! The conversation of using a proper polyfill has come up again in #20537, so I might imagine it could be a good place to consider again to incorporate `react-native-url-polyfill`.
